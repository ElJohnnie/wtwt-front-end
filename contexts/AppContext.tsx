import React, {
  createContext,
  useState,
  useRef,
  useCallback,
  ReactNode,
  FC,
} from 'react';
import { AppContextProps, Step, Questions } from '../types';

const defaultState = {
  steps: { current: [] as Step[] },
  currentStep: 0,
  questions: [],
  answeredSteps: [] as boolean[],
  answers: [] as string[],
  nextStep: () => {
    throw new Error('nextStep function not implemented');
  },
  prevStep: () => {
    throw new Error('prevStep function not implemented');
  },
  answerQuestion: () => {
    throw new Error('answerQuestion function not implemented');
  },
  showResult: false,
};

export const AppContext = createContext<AppContextProps>(defaultState);

interface AppProviderProps {
  children: ReactNode;
}

export const AppProvider: FC<AppProviderProps> = ({ children }) => {
  const steps = useRef<Step[]>([
    { number: '1', hasRange: true, isInactive: false, asComplete: false },
    { number: '2', hasRange: true, isInactive: true, asComplete: false },
    { number: '3', hasRange: true, isInactive: true, asComplete: false },
    { number: '4', hasRange: true, isInactive: true, asComplete: false },
    { number: '5', hasRange: false, isInactive: true, asComplete: false },
  ]);

  const questions: Questions[] = [
    {
      question: 'Como você está se sentindo agora?',
      answers: [
        { text: 'Feliz' },
        { text: 'Triste' },
        { text: 'Ansioso' },
        { text: 'Bem animado' },
        { text: 'Entediado' },
      ],
      description: 'teste 123',
    },
    {
      question: 'Qual o seu estilo de filme favorito?',
      answers: [
        { text: 'Ação' },
        { text: 'Aventura' },
        { text: 'Animação' },
        { text: 'Comédia' },
        { text: 'Drama' },
        { text: 'Fantasia' },
        { text: 'SciFi' },
        { text: 'Thriller' },
        { text: 'Terror/Suspense' },
      ],
      description: 'teste 123',
    },
    {
      question: 'Seu segundo estilo de filme favorito?',
      answers: [
        { text: 'Ação' },
        { text: 'Aventura' },
        { text: 'Animação' },
        { text: 'Comédia' },
        { text: 'Drama' },
        { text: 'Fantasia' },
        { text: 'SciFi' },
        { text: 'Thriller' },
        { text: 'Terror/Suspense' },
      ],
      description: 'teste 123',
    },
    {
      question: 'Qual periodo de lançamento mais lhe convém?',
      answers: [
        { text: '2000' },
        { text: '1990' },
        { text: '1980' },
        { text: '1970' },
        { text: '1960' },
        { text: '1950' },
        { text: '1940' },
      ],
      description: 'teste 123',
    },
  ];

  const [currentStep, setCurrentStep] = useState<number>(0);

  const [showResult, setShowResult] = useState<boolean>(false);

  const [answeredSteps, setAnsweredSteps] = useState<boolean[]>([
    false,
    false,
    false,
    false,
    false,
  ]);
  const [answers, setAnswers] = useState<string[]>([]);

  const nextStep = useCallback(() => {
    setCurrentStep((prev) => Math.min(prev + 1, steps.current.length - 1));
  }, []);

  const prevStep = useCallback(() => {
    setCurrentStep((prev) => Math.max(prev - 1, 0));
  }, []);

  const answerQuestion = useCallback(
    (text: string) => {
      if (currentStep < questions.length - 1) {
        setAnswers((prev) => {
          const newAnswers = [...prev];
          newAnswers[currentStep] = text;
          return newAnswers;
        });
        setAnsweredSteps((prev) => {
          const newAnsweredSteps = [...prev];
          newAnsweredSteps[currentStep] = true;
          return newAnsweredSteps;
        });
        nextStep();
      } else {
        setAnswers((prev) => {
          const newAnswers = [...prev];
          newAnswers[currentStep] = text;
          return newAnswers;
        });
        setAnsweredSteps((prev) => {
          const newAnsweredSteps = [...prev];
          newAnsweredSteps[currentStep] = true;
          return newAnsweredSteps;
        });
        setShowResult(true);
      }
    },
    [currentStep, nextStep, questions.length],
  );

  return (
    <AppContext.Provider
      value={{
        steps,
        currentStep,
        questions,
        answeredSteps,
        answers,
        nextStep,
        prevStep,
        answerQuestion,
        showResult,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
