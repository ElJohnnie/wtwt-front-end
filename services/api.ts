import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: process.env.BFF_BASE_URL,
  timeout: 5000,
  headers: {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${process.env.AUTHORIZATION_TOKEN} `,
  },
});

const fetchData = async (url: string, options = {}) => {
  try {
    const response = await axiosInstance(url, options);
    return response.data;
  } catch (error) {
    console.error('Error retrieving data:', error);
    throw new Error('Could not get data');
  }
};

export { fetchData };
