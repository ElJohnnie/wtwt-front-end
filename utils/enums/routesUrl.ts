export enum RoutesUrls {
  HOME = '/',
  RESULT = '/result',
  ERROR = '/error',
}

export type RouteUrl = RoutesUrls.RESULT | RoutesUrls.HOME | RoutesUrls.ERROR;
