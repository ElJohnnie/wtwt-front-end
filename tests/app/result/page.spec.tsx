import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import ResultPage from '../../../app/result/page';
import { AppContext } from '../../../contexts/AppContext';
import { axiosInstance } from '../../../services/api';
import { RoutesUrls } from '../../../utils/enums/routesUrl';

const redirectMock = jest.fn();
const replaceMock = jest.fn();

jest.mock('../../../hooks/useNavigation/index.tsx', () => ({
  ...jest.requireActual('../../../hooks/useNavigation/index.tsx'),
  useNavigation: () => ({
    redirect: redirectMock,
    replace: replaceMock,
  }),
}));

jest.mock('next/image', () => ({
  __esModule: true,
  default: (
    props: React.JSX.IntrinsicAttributes &
      React.ClassAttributes<HTMLImageElement> &
      React.ImgHTMLAttributes<HTMLImageElement>,
  ) => <img {...props} />,
}));

const mockContextValue = {
  steps: {
    current: [
      { number: '1', hasRange: true, isInactive: false, asComplete: false },
      { number: '2', hasRange: true, isInactive: true, asComplete: false },
      { number: '3', hasRange: true, isInactive: true, asComplete: false },
      { number: '4', hasRange: true, isInactive: true, asComplete: false },
      { number: '5', hasRange: false, isInactive: true, asComplete: false },
    ],
  },
  currentStep: 0,
  questions: [
    {
      question: 'Como você está se sentindo agora?',
      answers: [
        { text: 'Feliz', value: 'happy' },
        { text: 'Triste', value: 'sad' },
        { text: 'Ansioso', value: 'anxious' },
        { text: 'Bem animado', value: 'excited' },
        { text: 'Entediado', value: 'bored' },
      ],
      description: 'teste 123',
    },
  ],
  answeredSteps: [true, true, true, true],
  answers: ['any', 'any', 'any', 'any'],
  nextStep: jest.fn(),
  prevStep: jest.fn(),
  answerQuestion: jest.fn(),
  showResult: false,
  resetState: jest.fn(),
};

const mockMovie = {
  adult: false,
  backdrop_path: '/path/to/backdrop.jpg',
  genre_ids: [28, 12],
  id: 123,
  original_language: 'en',
  original_title: 'Mock Movie Title',
  overview: 'A brief overview of the movie.',
  popularity: 7.8,
  poster_path: '/path/to/poster.jpg',
  release_date: '2024-07-22',
  title: 'Mock Movie Title',
  video: false,
  vote_average: 8.1,
  vote_count: 1500,
};

const mockMoviesResponse = {
  recommendedMovie: {
    title: 'Recommended Movie Title',
  },
  detailedMovie: {
    page: 1,
    results: [mockMovie, mockMovie],
    total_pages: 10,
    total_results: 20,
  },
};

describe('ResultPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('Renderizando a página de resultado', async () => {
    jest
      .spyOn(axiosInstance, 'get')
      .mockResolvedValueOnce({ data: mockMoviesResponse });

    const { container } = render(
      <AppContext.Provider value={mockContextValue}>
        <ResultPage />
      </AppContext.Provider>,
    );

    await waitFor(() => {
      expect(screen.queryByRole('img')).toBeInTheDocument();
    });

    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('Clicando no botão de voltar para a página inicial', async () => {
    jest
      .spyOn(axiosInstance, 'get')
      .mockResolvedValueOnce({ data: mockMoviesResponse });

    render(
      <AppContext.Provider value={mockContextValue}>
        <ResultPage />
      </AppContext.Provider>,
    );

    await waitFor(() => {
      expect(screen.getByTestId('go-to-start')).toBeInTheDocument();
    });

    const backButton = screen.getByTestId('go-to-start');
    expect(backButton).toBeInTheDocument();

    fireEvent.click(backButton);

    expect(replaceMock).toHaveBeenCalledWith(RoutesUrls.HOME);
  });

  it('Impedindo de acessar a página caso não tenha quatro respostas', async () => {
    mockContextValue.answers = [];

    render(
      <AppContext.Provider value={mockContextValue}>
        <ResultPage />
      </AppContext.Provider>,
    );

    expect(replaceMock).toHaveBeenCalledWith(RoutesUrls.HOME);
  });
});
