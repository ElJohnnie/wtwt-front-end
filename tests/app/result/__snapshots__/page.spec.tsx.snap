// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultPage Renderizando a p√°gina de resultado 1`] = `
<div>
  <main
    class="flex min-h-screen flex-col items-center justify-center p-12"
  >
    <div
      class="flex flex-col items-center my-4"
    >
      <div
        class="flex flex-col justify-center items-center mx-auto"
      >
        <figure
          class="w-full"
        >
          <img
            alt="image description"
            height="80"
            layout="responsive"
            src="undefinedundefined"
            width="80"
          />
        </figure>
        <h3
          class="font-bold text-pink-600 text-center text-lg mt-3 max-w-x"
        />
        <p
          class="text-pink-600 text-center text-sm mt-2 max-w-sm"
        />
      </div>
    </div>
    <div
      class="my-4 grid"
    >
      <div
        class="w-full flex flex-col justify-center items-center"
      >
        <button
          class="w-24 h-24 rounded-full   inline-flex items-center justify-center   text-white text-[24px] font-bold bg-pink-600 hover:bg-pink-800"
          data-testid="go-to-start"
        >
          <svg
            class="size-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 9V5.25A2.25 2.25 0 0 1 10.5 3h6a2.25 2.25 0 0 1 2.25 2.25v13.5A2.25 2.25 0 0 1 16.5 21h-6a2.25 2.25 0 0 1-2.25-2.25V15m-3 0-3-3m0 0 3-3m-3 3H15"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
        <p
          class="text-[18px] mt-1 text-pink-600"
        >
          Refazer novamente
        </p>
      </div>
    </div>
  </main>
</div>
`;
